cpp/velox/shuffle/VeloxShuffleWriter.cc

```diff
@@ -167,8 +167,78 @@ std::shared_ptr<arrow::Array> makeBinaryArray(
+      totalSize += codec->MaxCompressedLen(buffer->size(), nullptr);
+    }
+  }
+  return totalSize;
+}
+
+std::shared_ptr<arrow::RecordBatch> makeCompressedRecordBatch(
+    uint32_t numRows,
+    const std::vector<std::shared_ptr<arrow::Buffer>>& buffers,
+    const std::shared_ptr<arrow::Schema> compressWriteSchema,
+    ShuffleBufferPool* pool,
+    arrow::util::Codec* codec) {
+  std::vector<std::shared_ptr<arrow::Array>> arrays;
+  // header col, numRows
+  {
+    std::shared_ptr<arrow::Buffer> headerBuffer;
+    GLUTEN_THROW_NOT_OK(pool->allocate(headerBuffer, sizeof(uint32_t)));
+    memcpy(headerBuffer->mutable_data(), &numRows, sizeof(uint32_t));
+    int32_t compressType = static_cast<int32_t>(codec->compression_type());
+    memcpy(headerBuffer->mutable_data() + sizeof(uint32_t), &compressType, sizeof(int32_t));
+    arrays.emplace_back(makeBinaryArray(compressWriteSchema->field(0)->type(), headerBuffer, pool));
+  }
+
+  // Length buffer layout |buffers.size()|buffer1 unCompressedLength|buffer1 compressedLength| buffer2...
+  std::shared_ptr<arrow::Buffer> lengthBuffer;
+  GLUTEN_THROW_NOT_OK(pool->allocate(lengthBuffer, (buffers.size() * 2 + 1) * sizeof(int64_t)));
+  int64_t offset = 0;
+  writeInt64(lengthBuffer, offset, buffers.size());
+
+  int64_t compressedBufferMaxSize = getMaxCompressedBufferSize(buffers, codec);
+  std::shared_ptr<arrow::ResizableBuffer> valueBuffer;
+  GLUTEN_THROW_NOT_OK(pool->allocateDirectly(valueBuffer, compressedBufferMaxSize));
+  int64_t compressValueOffset = 0;
+  for (auto& buffer : buffers) {
+    if (buffer != nullptr && buffer->size() != 0) {
+      writeInt64(lengthBuffer, offset, buffer->size());
+      int64_t actualLength;
+      int64_t maxLength = codec->MaxCompressedLen(buffer->size(), nullptr);
+      GLUTEN_ASSIGN_OR_THROW(
+          actualLength,
+          codec->Compress(
+              buffer->size(), buffer->data(), maxLength, valueBuffer->mutable_data() + compressValueOffset));
+      compressValueOffset += actualLength;
+      writeInt64(lengthBuffer, offset, actualLength);
+      // std::cout << "before compress " << buffer->size() << "actualLength " << actualLength << std::endl;
+    } else {
+      writeInt64(lengthBuffer, offset, 0);
+      writeInt64(lengthBuffer, offset, 0);
+      // std::cout << "before compress 0 actualLength 0" << std::endl;
+    }
+  }
+  // std::cout << "writer buffer size " << buffers.size() << std::endl;
+  GLUTEN_THROW_NOT_OK(valueBuffer->Resize(compressValueOffset, /*shrink*/ true));
+
+  arrays.emplace_back(makeBinaryArray(compressWriteSchema->field(1)->type(), lengthBuffer, pool));
+  arrays.emplace_back(makeBinaryArray(compressWriteSchema->field(2)->type(), valueBuffer, pool));
+  return arrow::RecordBatch::Make(compressWriteSchema, 1, {arrays});
+}
+
 // generate the new big one row several columns binary recordbatch
-std::shared_ptr<arrow::RecordBatch> makeRecordBatch(
+std::shared_ptr<arrow::RecordBatch> makeUncompressedRecordBatch(
```

Comment:
We should copy the uncompressed buffer for perferSpill=False

Reply:
Yes, I will remove isTinyBatch only in next commit, left codec == nullptr check